Userland/Libraries/LibVideo/DecoderError.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2022, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/Format.h>
+#include <AK/String.h>
+#include <errno.h>
+
+namespace Video {
+
+struct DecoderError;
+
+template<typename T>
+using DecoderErrorOr = ErrorOr<T, DecoderError>;
+
+enum class DecoderErrorCategory : u32 {
+    Unknown = 0,
+    IO,
+    // The input is corrupted.
+    Corrupted,
+    // The input uses features that are not yet implemented.
+    NotImplemented,
+};
+
+struct DecoderError {
+public:
+    static DecoderError with_description(DecoderErrorCategory category, StringView description)
+    {
+        return DecoderError(category, description);
+    }
+
+    template<typename... Parameters>
+    static DecoderError format(DecoderErrorCategory category, CheckedFormatString<Parameters...>&& format_string, Parameters const&... parameters)
+    {
+        AK::VariadicFormatParams variadic_format_params { parameters... };
+        return DecoderError::with_description(category, String::vformatted(format_string.view(), variadic_format_params));
+    }
+
+    DecoderErrorCategory category()
+    {
+        return m_category;
+    }
+
+    StringView description()
+    {
+        return m_description;
+    }
+
+    StringView string_literal()
+    {
+        return m_description;
+    }
+
+    DecoderError with_caller(String caller)
+    {
+        return DecoderError::format(m_category, "{}: {}", caller, m_description);
+    }
+
+private:
+    DecoderError(DecoderErrorCategory category, String description)
+        : m_category(category)
+        , m_description(move(description))
+    {
+    }
+
+    DecoderErrorCategory m_category { DecoderErrorCategory::Unknown };
+    String m_description { String::empty() };
+};
+
+// VSCode doesn't know __FILE_NAME__ exists
+#ifndef __FILE_NAME__
+#    define __FILE_NAME__ ""
+#endif
```

Comment:
Is this code here just to make your IDE happy? If it builds without this, please remove it.

Reply:
Yeah, without it I had errors in every TRY statement in Parser.cpp. I've changed __FILE_NAME__ usage to __FILE__ instead, though that will be a lot more verbose.

Reply:
Sorry to comment on a closed item, but looking at the code here, I think you should embrace our lord and savior AK::SourceLocation and its magic SourceLocation::current() value instead of relying on these macros to create descriptive errors.